### PR TITLE
Write parsetable ATerms directly to file without making a string first

### PR DESCRIPTION
**Pull request by Apanatshka**
Also requires metaborg/spoofax#58 to be merged at the same time.

### CODE REVIEW

File: org.metaborg.parsetable/src/main/java/org/metaborg/parsetable/characterclasses/CharacterClassSingle.java [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 import org.spoofax.interpreter.terms.IStrategoTerm;
 import org.spoofax.interpreter.terms.ITermFactory;
+import org.spoofax.terms.util.B;
```
Comment (mpsijm) [inline org.metaborg.parsetable/src/main/java/org/metaborg/parsetable/characterclasses/CharacterClassSingle.java]: Can you also remove the unused imports? :innocent:

Comment (Apanatshka) [reply org.metaborg.parsetable/src/main/java/org/metaborg/parsetable/characterclasses/CharacterClassSingle.java]: Done

**Comment by Apanatshka** [comment]
@mpsijm Can you have another look at this PR? I just rebased the branch onto master. I was careful with manually resolving conflicts, but you never know if I made a mistake or Git made a mistake automatically.

File: org.metaborg.sdf2table/src/main/java/org/metaborg/sdf2table/grammar/CharacterClass.java [hunk 2]
```diff
@@ -82,8 +81,7 @@ public CharacterClass(BitSet chars, boolean eof) {
     }
 
     @Override public IStrategoTerm toAtermList(ITermFactory tf) {
-        List<IStrategoTerm> terms = new ArrayList<>();
-
+        IStrategoList.Builder terms = tf.arrayListBuilder(chars.size());
```
Comment (mpsijm) [inline org.metaborg.sdf2table/src/main/java/org/metaborg/sdf2table/grammar/CharacterClass.java]: For the full character class range (0-1114111), `chars.size() == 1114112` while there is only one range. This means that the backing array will be _very_ sparse. Can the initial size of the `arrayListBuilder` be left out here?

File: org.metaborg.sdf2table/src/main/java/org/metaborg/sdf2table/io/ParseTableIO.java [hunk 3]
```diff
@@ -267,9 +266,9 @@ private static IStrategoTerm generatePrioritiesAterm(ParseTable pt) throws Excep
     }
 
     private static IStrategoTerm generateLabelsAterm(ParseTable pt) {
-        List<IStrategoTerm> terms = Lists.newArrayList();
+        IStrategoList.Builder terms = termFactory.arrayListBuilder(pt.productionLabels().size());
 
-        for(int i = 257 + pt.productionLabels().size() - 1; i >= 257; i--) {
+        for(int i = 257 + pt.productionLabels().size() - 1, j = 0; i >= 257; i--, j++) {
```
[the remark this answers is not in the document]
Comment (Apanatshka) [reply org.metaborg.sdf2table/src/main/java/org/metaborg/sdf2table/io/ParseTableIO.java]: Nice catch, that should have been removed along with the other code that used to use it.

**Comment by Apanatshka** [comment]
@udesou Let's land this PR and https://github.com/metaborg/spoofax/pull/58 soon. I get the impression (by running it locally) that this speeds up the Java version of sdf2table quite a bit.

**Comment by udesou** [comment]
If it has no conflicts with the other changes, it should be good to go, afaik.

**Pull request merged**